compiletest: improve robustness of LLVM version handling

jieyouxu opened the pull request:
Previously, `extract_llvm_versions` did some gymnastics for llvm versions by combining `(major, minor, patch)` into a combined version integer, but that is not very robust and made it difficult to add `max-llvm-major-version`. This PR tries to:

- Improve llvm version handling robustness by parsing and representing the version as a semver. We intentionally deviate from strict semver standards by allowing omission of minor and patch versions. They default to `0` when absent. This is for convenience to allow the user to write e.g. `//@ min-llvm-version: 18` instead of having to spell out the full `major.minor.patch` semver string `//@ min-llvm-verison: 18.0.0`.
- Adjust some panic messages to include a bit more context about *why* the version string was rejected.

Prerequisite for #132310.

r? bootstrap (or compiler)

jieyouxu committed:
compiletest: improve robustness of LLVM version handling

jieyouxu commented on src/tools/compiletest/src/header.rs:
```diff
@@ -1113,34 +1113,45 @@ fn parse_normalize_rule(header: &str) -> Option<(String, String)> {
-        .map(|s| s.parse().expect("Malformed version component"))
+        .map(|s| s.parse().expect("llvm version component should consist of only digits"))
         .collect();
-    let version = match *components {
-        [a] => a * 10_000,
-        [a, b] => a * 10_000 + b * 100,
-        [a, b, c] => a * 10_000 + b * 100 + c,
-        _ => panic!("Malformed version"),
-    };
```
Specifically, this

onur-ozkan commented on src/tools/compiletest/src/header.rs:
```diff
@@ -1113,34 +1113,45 @@ fn parse_normalize_rule(header: &str) -> Option<(String, String)> {
+/// Currently panics if the input string is malformed, though we really should not use panic as an
+/// error handling strategy.
+///
+/// FIXME(jieyouxu): improve error handling
```
Did you consider using `semver`?

jieyouxu replied:
I did not immediately, since it wasn't in the dependencies for compiletest specifically, although we certainly could as I imagine `semver` is already in the workspace dep anyway.

jieyouxu replied:
I tried this locally, apparently `semver` adheres to "strict" semantic versioning, meaning that you cannot omit patch and minor versions if you want to parse the version string as "strictly" `semver::Version` (for good reasons).

i.e. this will be rejected, because `19` is not a valid semver (neither is `19.2`, but `19.0.0` is okay):

```
//@ min-llvm-version: 19
```

We *can* facade `semver`'s parsing logic and just stub out zeros, but it might be easier just to hand-write the parsing in the first place...

onur-ozkan replied:
> We can facade semver's parsing logic and just stub out zeros, but it might be easier just to hand-write the parsing in the first place... What do you think?

What about using zeros for the missing fields? Can't we treat "19" as "19.0.0"? This might be easier to handle and maintain.

jieyouxu replied:
I can do that.

jieyouxu replied:
Changed to use `semver` as suggested.

jieyouxu commented:
@rustbot author

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11575375085/job/32221830884) [(plain)](https://github.com/rust-lang/rust/commit/8b92efbfb6220272eeb24dcf3353f33ca949a786/checks/32221830884/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

jieyouxu commented:
Changes since last review:

- Use `semver` as suggested for handling the version, dropping hand-rolled `LlvmVersion`-related logic. The extraction logic facades over `semver::Version` parsing to permit omitting minor and patch version components.
- Obviously, this means that adding `semver` as a dependency to compiletest, but `semver` was already in workspace deps anyway.

onur-ozkan reviewed (approved):
LGTM - Waiting for CI.

onur-ozkan commented:
@bors r+
